### PR TITLE
Collapse foreach loops

### PR DESCRIPTION
**Pull request by rymrg**

### CODE REVIEW

File: source/schlib/lookup.d [hunk 1]
```diff
@@ -50,40 +50,32 @@ private struct LookupTable(K, V, bool useTruthiness)
...
+            immutable hash = hashOf(key);
+            for (size_t i = 0, idx = hash % buckets.length; i < buckets.length; ++i, idx = idx + 1 % buckets.length)
+            {
+                auto b = &(buckets[idx]);
+                if(!*b)
                     // this is where it would have gone...
                     return null;
                 else if(b.hash == hash && b.key == key)
                     return &b.value;
+            }
```
Comment (schveiguy) [inline source/schlib/lookup.d]: Hm.. this is doing a division on every loop.

I suggest this (just thought of it this morning):

```d
import std.range : only;
foreach(arr; only(buckets[idx .. $], buckets[0 .. idx]))
    foreach(ref b; arr)
          ....
```

Comment (rymrg) [reply source/schlib/lookup.d]: I tried to benchmark the various options (with a contrived micro benchmark). It seems you're right and the division does not get optimized nicely.
On dmd, `only` seems to be fastest and beat `double for` by a small margin and chain is as slow as the `modulus`.
On the other hand, on ldc, chain seems to be the fastest option but `only` and `double for` are not far behind.

**Comment by rymrg** [comment]
inout doesn't seem to play nicely with only. Any ideas?

**Comment by schveiguy** [comment]
> inout doesn't seem to play nicely with only. Any ideas?

Grrrr... That's unfortunate. I wish I could fix that problem...

Another possibility is to do it a bit more clunkily:

```d
foreach(x; 0 .. 2)
   foreach(ref b; x ? buckets[idx .. $] : buckets[0 .. idx])
      ...
```

Should be reasonably fast, and kind of match what `only` would do.

**Comment by rymrg** [comment]
> Should be reasonably fast, and kind of match what `only` would do.
Seems like it's much slower. Not has bad as %, but noticeable.

**Comment by schveiguy** [comment]
> Seems like it's much slower. Not has bad as %, but noticeable.

You mean the outer for loop is slower? Strange. You can change it to static foreach, and it should be just like you wrote the original 2 loops.

**Comment by schveiguy** [comment]
Oh wait, I got my condition BACKWARDS. That is not valid.

**Comment by schveiguy** [comment]
Should be:
```d
static foreach(const x; 0 .. 2)
    foreach(ref b; x ? buckets[0 .. idx] : buckets[idx .. $])
```

**Comment by schveiguy** [comment]
Yeah, go with the static foreach. That should be identical to the original code, while not having to repeat the loop.

**Comment by rymrg** [comment]
> Oh wait, I got my condition BACKWARDS. That is not valid.

Valid but slower. Yeah, that explains the margin.

> Yeah, go with the static foreach. That should be identical to the original code, while not having to repeat the loop.

Probably simplest. Should I keep the only or make both static foreach for consistency?

**Comment by schveiguy** [comment]
Use static foreach for both. I'd prefer less dependencies.

**Comment by schveiguy** [comment]
> Valid but slower.

Not so much for a hash table. What was happening is it was considering everything to be hashed to bucket index 0, and then just doing a linear search through the array to find an empty spot ;) Basically, that replaced the hash table with an array!

**Comment by rymrg** [comment]
> > Valid but slower.
> 
> Not so much for a hash table. What was happening is it was considering everything to be hashed to bucket index 0, and then just doing a linear search through the array to find an empty spot ;) Basically, that replaced the hash table with an array!

Do you consider O notation part of correctness? :)

**Comment by schveiguy** [comment]
> Do you consider O notation part of correctness? :)

That's the primary thing that matters if you are doing a hash table! If you want to use an array, use an array.

"Always start at bucket 0" is O(n) lookup. But yes, it would pass unit tests still ;) Those are the kinds of things that are really hard to test for. (I could probably do a unittest with average collisions per item, and verify it's below a certain number).

**Comment by schveiguy** [comment]
Thanks for the PR! I'll tag too so you can use it.

**Pull request merged**